### ISSUE TITLE
REST API: if invalid schema is provided server returns HTTP status code 404

### ISSUE THREAD

**Issue by pansapiens** [mytardis/mytardis#440]
If an invalid schema is provided for a DataFile upload via HTTP POST the API currently returns HTTP status code 404 (Not Found). This can be confusing when writing a client since the URL endpoint itself exists and may have succeeded in past request, but only fails with 404 depending on the content of the JSON message body (ie if the schema is known to the MyTardis instance or not). A more appropriate status code in this case would be 400 (Bad Request), which implies that the URL endpoint exists but the request content is malformed or invalid in some way.

This may also apply to creation of Experiments and Datasets using the REST API.

**Comment by wettenhj** [comment]
If MyTardis is running in DEBUG mode, then you can get more info than just a 404 - check out the 'error_message' part of the response content.  For example, when creating an experiment via the API, MyData sometimes encounters a 404 error because of a missing Schema, in which the 'error_message' returned by the MyTardis API is "Schema matching query does not exist.".  But another possibility is a missing UserProfile record, in which case the 'error_message' is "UserProfile matching query does not exist.".  See: https://github.com/monash-merc/mydata/blob/develop/mydata/ExperimentModel.py#L73

**Comment by wettenhj** [comment]
So I guess I disagree with your assertion that this type of problem implies that "the request content is malformed or invalid".  There could be something misconfigured on the server-side, e.g. a User record created without a corresponding UserProfile record, which doesn't relate to the "request content" of the API query.

**Comment by pansapiens** [comment]
@wettenhj Yes, the associated error messages are informative, which is how I ultimately discovered *why* I was getting an error. However, in HTTP REST terms 404 should only be used if the request URI doesn't exist, not in cases where the resource exists but can't be returned due to something in the body of the message (ie, http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html only mentions the Request-URI, not the POST body). 

I don't think status code 400 is *technically* quite right either - the RFC says this is when the server encounters malformed syntax - syntactically correct JSON with an unknown schema value doesn't really fulfil this requirement. However, consensus usage across various popular web APIs (eg Twitter, Imgur, Reddit, Dropbox, Uniprot) suggests that if a parameter is out of range or invalid (even with valid syntax), 400 is the status code typically returned.

Not sure about the specifics of the User / UserProfile example, but it sounds like it could be more appropriate to return a 500 (Internal Server Error) in this case - the request was entirely valid (User exists) but some issue with the internal state of the server (missing associated UserProfile) means it couldn't be fulfilled.

It's a minor issue but it probably should be addressed in the future as part of the transition to API v2 and Django REST Framework.

**Comment by wettenhj** [comment]
@pansapiens, I have no problem with API v2 behaving differently with regard to 404s and other status codes.  And I certainly wasn't trying to argue that TastyPie's HTTP status codes (as used in v1) are perfect.  I agree that a 500 (Internal Server Error) would seem more appropriate for a missing UserProfile record.  My main reason for commenting was to point out that if the status code returned by API v1 for a missing schema was changed, then this would affect existing MyData code.
